In [1]:
import graphlab

In [5]:
song_data = graphlab.SFrame('song_data.gl/')
graphlab.canvas.set_target('ipynb')
song_data.print_rows(num_rows=3)
print 'len of song_data = ', len(song_data)
print 'len of unique user id= ', len(song_data['user_id'].unique())
print 'len of unique song = ', len(song_data['song'].unique())

+-------------------------------+--------------------+--------------+
|            user_id            |      song_id       | listen_count |
+-------------------------------+--------------------+--------------+
| b80344d063b5ccb3212f76538f... | SOAKIMP12A8C130995 |      1       |
| b80344d063b5ccb3212f76538f... | SOBBMDR12A8C13253B |      2       |
| b80344d063b5ccb3212f76538f... | SOBXHDL12A81C204C0 |      1       |
+-------------------------------+--------------------+--------------+
+-----------------+---------------+-------------------------------+
|      title      |     artist    |              song             |
+-----------------+---------------+-------------------------------+
|     The Cove    |  Jack Johnson |    The Cove - Jack Johnson    |
| Entre Dos Aguas | Paco De Lucia | Entre Dos Aguas - Paco De ... |
|     Stronger    |   Kanye West  |     Stronger - Kanye West     |
+-----------------+---------------+-------------------------------+
[1116609 rows x 6 columns]

len of

In [6]:
## question 1: find out the # of unique users for each of the artist: Kanye West, Foo FIghters, Taylor Swift, Lady GaGa

In [7]:
artist = ['Kanye West', 'Foo Fighters', 'Taylor Swift', 'Lady GaGa']

In [8]:
for i in artist:
    user_artist = song_data[song_data['artist'] == i]
    print 'for ', i, ' , # of unique users are = ', len(user_artist['user_id'].unique())

for  Kanye West  , # of unique users are =  2522
for  Foo Fighters  , # of unique users are =  2055
for  Taylor Swift  , # of unique users are =  3246
for  Lady GaGa  , # of unique users are =  2928


# Question 1:
# Kayne West, num of unique users are: 2522
# Foo Fighters, 2055
# Taylor Swift: 3246
# Lady Gaga: 2928

In [10]:
## Question 2: find out the most popular and least popular artist

In [15]:
artist_count = song_data.groupby(key_columns='artist', 
                                 operations={'total_count': graphlab.aggregate.SUM('listen_count')})
artist_count = artist_count.sort('total_count', ascending=False)
print artist_count[0:5]
artist_count.tail()

+------------------------+-------------+
|         artist         | total_count |
+------------------------+-------------+
|     Kings Of Leon      |    43218    |
|     Dwight Yoakam      |    40619    |
|         Björk          |    38889    |
|        Coldplay        |    35362    |
| Florence + The Machine |    33387    |
+------------------------+-------------+
[5 rows x 2 columns]



Columns:
	artist	str
	total_count	int

Rows: 10

Data:
+-------------------------------+-------------+
|             artist            | total_count |
+-------------------------------+-------------+
|          Jody Bernal          |      38     |
|        Aneta Langerova        |      38     |
| Kanye West / Talib Kweli /... |      38     |
|             Nâdiya            |      36     |
|         harvey summers        |      31     |
|         Boggle Karaoke        |      30     |
|             Diplo             |      30     |
| Beyoncé feat. Bun B and Sl... |      26     |
|         Reel Feelings         |      24     |
|        William Tabbert        |      14     |
+-------------------------------+-------------+
[10 rows x 2 columns]

# Question 2: 
# The most popular artist = Kings ofLeon: 43218
# The least artist: William Tabbert: 14

In [17]:
## Question 3: build the recommender

In [18]:
train_data, test_data = song_data.random_split(.8, seed=0)

In [19]:
subset_test_users = test_data['user_id'].unique()[0:10000]

In [ ]:
personalized_model = graphlab.